src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.TranslationAction_Actions.cs

```diff
@@ -62,12 +72,14 @@ internal sealed class TouchAdditionalDocumentAction(AdditionalDocumentState oldS
                 public override Task<Compilation> TransformCompilationAsync(Compilation oldCompilation, CancellationToken cancellationToken)
                     => Task.FromResult(oldCompilation);
 
-                public override CompilationAndGeneratorDriverTranslationAction? TryMergeWithPrior(CompilationAndGeneratorDriverTranslationAction priorAction)
+                public override TranslationAction? TryMergeWithPrior(TranslationAction priorAction)
                 {
                     if (priorAction is TouchAdditionalDocumentAction priorTouchAction &&
                         priorTouchAction._newState == _oldState)
                     {
-                        return new TouchAdditionalDocumentAction(priorTouchAction._oldState, _newState);
+                        // As we're merging ourselves with the prior touch action, we want to keep the old project state
+                        // that we are translating from.
+                        return new TouchAdditionalDocumentAction(priorAction.OldProjectState, priorTouchAction._oldState, _newState);
```

Comment:
this preserves the sematnics that after merging, our 'oldstate' was the original old state of the prior action.